Fix Backend Cache-Clearing With Custom Mount Path

DanielWright opened the pull request:
When Spree is mounted to anything other than '/', the "Clear Cache" button in the General Settings page ceases working, because the event-handler path is hard-coded.

This fix addresses this by refactoring the cache-clearing path into a data-attribute in the view-template, which is retrieved dynamically by the event-handler.

The included spec is the best I could do to exercise this edge case. Let me know if you have a more elegant way to remount Spree for this example.

Commit:
Fix Backend Cache-Clearing With Custom Mount Path

When Spree is mounted to anything other than '/', the "Clear Cache"
button in the General Settings page ceases working, because the
event-handler path is hard-coded.

This fix addresses this by refactoring the cache-clearing path into a
data-attribute in the view-template, which is retrieved dynamically by
the event-handler.

The included spec is the best I could do to exercise this edge case. Let
me know if you have a more elegant way to remount Spree for this
example.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://danielwright-fix-cache-flush-with-custom-engine-path.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://danielwright-fix-cache-flush-with-custom-engine-path.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
Delinting (h/t @houndci)

JDutil commented on backend/app/assets/javascripts/spree/backend/general_settings.js.coffee:
```diff
@@ -1,8 +1,8 @@
     $.ajax
       type: 'POST'
-      url: '/admin/general_settings/clear_cache'
+      url: (($ event.target).attr 'data-url')
```
Could you update this so that instead you set a route here:
https://github.com/spree/spree/blob/master/backend/app/assets/javascripts/spree/backend.js#L16-L36

Then can just call the route like: Spree.routes.clear_cache

JDutil commented on backend/spec/features/admin/configuration/general_settings_spec.rb:
```diff
@@ -39,5 +39,31 @@
       expect(page).to have_content(Spree.t(:clear_cache_ok))
     end
+
+    context "with Spree mounted to a custom path" do
```
I don't think we need to add another spec just for this on the already slow backend.  The hardcoded url should have been caught during code review.

DanielWright replied:
Okay, I'll remove the spec. (Darn, and after I spent a bunch of time figuring out how to do this! :smile: )

Commit:
Removes Unnecessary Spec

Commit:
Refactor Event-Handler Spree Route for Consistency

In my original PR, my implementation of mount-agnostic routing was
inconsistent with the established convention. This oversight should now
be resolved.

DanielWright commented on backend/app/assets/javascripts/spree/backend.js:
```diff
@@ -13,6 +13,7 @@
 //= require spree/backend/spree-select2
 //= require_tree .
 
+Spree.routes.clear_cache = Spree.pathFor('admin/general_settings/clear_cache')
```
Not touching, as this seems consistent with the following LOCs.

Commit:
Removes Now-Unnecessary Data Attributes

DanielWright commented:
Will this be backported into `2-4-stable` (and other relevant stable branches), or should I submit a separate PR?

JDutil commented:
Done

DanielWright commented:
Amazing, thanks! :+1:
